lib/bandit/http1/handler.ex

```diff
@@ -54,19 +47,73 @@ defmodule Bandit.HTTP1.Handler do
     end
   end
 
-  # Build URI dependent on request target type
-  defp build_uri(scheme, host, {:abs_path, path}),
-    do: URI.parse("#{scheme}://#{host}#{path}")
+  defp determine_scheme({:absoluteURI, scheme, _, _, _}, req) do
+    case {Adapter.secure?(req), scheme} do
+      {true, :https} -> {:ok, "https"}
+      {false, :http} -> {:ok, "http"}
+      _ -> {:error, "request target scheme does not agree with transport"}
+    end
+  end
 
-  defp build_uri(_scheme, _host, {:absoluteURI, scheme, host, :undefined, path}),
-    do: URI.parse("#{scheme}://#{host}#{path}")
+  defp determine_scheme(_request_target, req) do
+    if Adapter.secure?(req), do: {:ok, "https"}, else: {:ok, "http"}
+  end
 
-  defp build_uri(_scheme, _host, {:absoluteURI, scheme, host, port, path}),
-    do: URI.parse("#{scheme}://#{host}:#{port}#{path}")
+  defp determine_host_and_port({:absoluteURI, _, host, port, _}, _headers, req) do
+    case port do
+      :undefined -> {:ok, to_string(host), Adapter.get_local_data(req)[:port]}
+      port -> {:ok, to_string(host), port}
+    end
+  end
 
-  defp build_uri(scheme, host, :*) do
-    URI.parse("#{scheme}://#{host}/*")
-    |> Map.put(:path, "*")
+  defp determine_host_and_port(_request_target, headers, req) do
+    with host_header when is_binary(host_header) <- Adapter.get_header(headers, "host"),
+         {:ok, host, port} <- parse_host_header(host_header) do
+      case port do
+        :undefined -> {:ok, host, Adapter.get_local_data(req)[:port]}
+        port -> {:ok, host, port}
+      end
+    else
+      nil ->
+        case req.version do
+          :"HTTP/1.0" -> {:ok, "", Adapter.get_local_data(req)[:port]}
+          _ -> {:error, "No host header"}
+        end
+
+      error ->
+        error
+    end
+  end
```

Comment:
Not important at all, but this might be a little more obvious which errors come from where (if I read it correctly)

```suggestion
  defp determine_host_and_port(_request_target, headers, req) do
    case Adapter.get_header(headers, "host") do
      nil ->
        case req.version do
          :"HTTP/1.0" -> {:ok, "", Adapter.get_local_data(req)[:port]}
          _ -> {:error, "No host header"}
        end
        
      host_header ->
        with {:ok, host, port} <- parse_host_header(host_header) do
          case port do
            :undefined -> {:ok, host, Adapter.get_local_data(req)[:port]}
            port -> {:ok, host, port}
          end
        end
    end
  end
```

Reply:
IIRC I had it laid out that way (or substantially that way) originally, and credo was complaining about the cyclomatic complexity of the function being too high.

Reply:
I agree that some of these functions can be a bit tricky to read though!